lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -583,6 +584,265 @@ class ConvertAtenMaxPool2dWithIndicesOp
+          loc, rewriter.getZeroAttr(selfType.getElementType()));
+      Value invalidIdx = rewriter.create<arith::ConstantOp>(
+          loc, rewriter.getIntegerAttr(indicesType.getElementType(), -1));
+      self = createPad(self, zero, low, high);
+      indices = createPad(indices, invalidIdx, low, high);
+    }
+
+    Value init = rewriter.create<tensor::EmptyOp>(
+        loc, getAsOpFoldResult(outSizePadded), selfType.getElementType());
+
+    SmallVector<AffineExpr> inputExprs;
+    SmallVector<AffineExpr> outputExprs;
+    for (auto i : llvm::seq<int64_t>(0, outRank)) {
+      AffineExpr dim = rewriter.getAffineDimExpr(i);
+      if (i < NC) {
+        inputExprs.emplace_back(dim);
+      } else {
+        int64_t j = i - NC;
+        inputExprs.emplace_back(dim.floorDiv(stride[j]));
+      }
+      outputExprs.emplace_back(dim);
+    }
+
+    SmallVector<AffineMap> indexingMaps = AffineMap::inferFromExprList(
+        {inputExprs, inputExprs, outputExprs}, rewriter.getContext());
+
+    SmallVector<utils::IteratorType> iteratorTypes(
+        outRank, utils::IteratorType::parallel);
+
+    auto computeIndex = [&](OpBuilder &b, Location loc) -> Value {
+      // Next linalg.generic uses identity mapping for the unpooled tensor,
+      // compute linear index for output element, which we will the compare with
+      // values which came from indices tensor.
+      Value ret;
+      for (auto i : llvm::seq<int64_t>(NC, outRank)) {
+        Value idx = b.create<linalg::IndexOp>(loc, i);
+        // If pool input was padded, adjust indices so they start at 0 in the
+        // non-padded area. Indices outside non-padded area will make no sense,
+        // but it doesnt matter as we will cut the padded area later by
+        // extract_slice.
+        int64_t pad = padding[i - NC];
+        if (pad != 0) {
+          Value padVal = b.create<arith::ConstantIndexOp>(loc, pad);
+          idx = b.create<arith::SubIOp>(loc, idx, padVal);
+        }
+
+        if (!ret) {
+          ret = idx;
+        } else {
+          Value size =
+              b.create<arith::ConstantIndexOp>(loc, resType.getShape()[i]);
+          ret = b.create<arith::MulIOp>(loc, ret, size);
+          ret = b.create<arith::AddIOp>(loc, ret, idx);
+        }
+      }
+      return ret;
+    };
+
+    auto builder = [&](OpBuilder &b, Location loc, ValueRange args) {
+      // Compute current output linear index and compare it with the value
+      // from indices arg.
+      Value input = args[0];
+      Value zero = b.create<arith::ConstantOp>(
+          loc, rewriter.getZeroAttr(input.getType()));
+      Value index = b.create<arith::IndexCastOp>(loc, indexType, args[1]);
+      Value currentIndex = computeIndex(b, loc);
+      Value cmp = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq, index,
+                                          currentIndex);
+      Value out = b.create<arith::SelectOp>(loc, cmp, input, zero);
+      b.create<linalg::YieldOp>(loc, out);
```

Comment:
Okay. 

So given linalg indices, i, you compute the "flattened index" of i within the generic output (unpool tensor) with `computeIndex`. i is used to access the smaller `self` and `indices` through a floor divide with the stride, and `indices[i//stride]` is compared to `computeIndex`. Under what circumstances would a given `indices[i//stride]` not be equal to any `computeIndex` values for corresponding `i`'s? I suppose if the kernel size is greater than the stride, then this fails like you mention in the initial comment, since the value from `index[i//stride]` may have come from something outside of `[(i//stride)*stride, (i//stride)*stride + 1)` from the unpool tensor, which is the only region where you are checking for a match. 

I think this approach is probably reasonably fast, since you are functionally able to minimize the looping to the size of the unpool tensor (and don't need arbitrary read/write), but definitely has a drawback of not being robust to index values coming from larger kernel sizes.  

On the other hand, writing a loop (scf or linalg generic) over the indices and self tensors and then using `tensor::insert` or something like that would be more robust, but may come at a performance loss due to the arbitrary memory access? 

Do I understand the tradeoffs correctly?

Reply:
Some points:
* Loop (`scf.for`) implementation will be much more mentally simpler, but I'm not sure actual pipeline will be able to lower it efficiently (can it parallelize it? lower for gpu?). It will need slightly more memory accesses, as you will need to fill the entire output tensor with zeros first and then overwrite some values. Also, I'm pretty sure direct `tensor.insert` is not allowed inside `linalg.generic` (I see a few places where we have `tensor.extract` inside generic body, which is already dubious thing to do, IMO). 
* This approach can be extended for overlapping windows, but it will require even more linalg gymnastics and performance implications much less clear.
For non-overlapping windows we are doing mapping like this (1d case for simplicity, indices here and later - indices on which we access input and output tensors):
```
in_size=2, out_size=4, kernel=2, stride=2
self indices: [0, 0, 1, 1]
out indices:  [0, 1, 2, 3]
```
For overlapping windows, input value may affect multiple output values, so we will have to do mapping like this (note that linalg output is 2x bigger than actual output):
```
in_size=3, out_size=4, kernel=2, stride=1
self indices: [0, 0, 0, 1, 1, 2, 2, 2]
out indices:  [0, 0, 1, 1, 2, 2, 3, 3]
```
And then we can do a reduction on output, reducing size by 2x and taking non-zero values.
I had a rough draft for this approach, but I was stuck without knowing kernel size.
* None of this will work for `max_unpool2d` (but direct loop will) as the only info we have is output size, but onnx ops actually seems to preserve all necessary info, so direct onnx->linalg lowering might be possible.